--- a/arch/arm/boot/dts/bcm2711-rpi-cm4.dts 2022-01-04 02:51:16.000000000 +0800
+++ b/arch/arm/boot/dts/bcm2711-rpi-cm4.dts    2022-01-04 03:10:34.000000000 +0800
@@ -91,19 +91,12 @@
 		gpio-line-names = "BT_ON",
 				  "WL_ON",
 				  "PWR_LED_OFF",
-				  "ANT1",
 				  "VDD_SD_IO_SEL",
 				  "CAM_GPIO",
 				  "SD_PWR_ON",
 				  "ANT2";
 		status = "okay";
 
-		ant1: ant1 {
-			gpio-hog;
-			gpios = <3 GPIO_ACTIVE_HIGH>;
-			output-high;
-		};
-
 		ant2: ant2 {
 			gpio-hog;
 			gpios = <7 GPIO_ACTIVE_HIGH>;
@@ -138,23 +131,10 @@
 			  "SPI_MISO",
 			  "SPI_MOSI",
 			  "SPI_SCLK",
-			  "GPIO12",
-			  "GPIO13",
 			  /* Serial port */
 			  "TXD1",
 			  "RXD1",
-			  "GPIO16",
-			  "GPIO17",
-			  "GPIO18",
-			  "GPIO19",
-			  "GPIO20",
-			  "GPIO21",
-			  "GPIO22",
-			  "GPIO23",
-			  "GPIO24",
 			  "GPIO25",
-			  "GPIO26",
-			  "GPIO27",
 			  "RGMII_MDIO",
 			  "RGMIO_MDC",
 			  /* Used by BT module */
@@ -368,14 +348,9 @@
 		mmc2 = &sdhost;
 		i2c3 = &i2c3;
 		i2c4 = &i2c4;
-		i2c5 = &i2c5;
-		i2c6 = &i2c6;
 		i2c20 = &ddc0;
 		i2c21 = &ddc1;
-		spi3 = &spi3;
 		spi4 = &spi4;
-		spi5 = &spi5;
-		spi6 = &spi6;
 		/delete-property/ intc;
 	};
 
@@ -398,49 +373,7 @@
 	pinctrl-0 = <&uart1_pins>;
 };
 
-&spi0 {
-	pinctrl-names = "default";
-	pinctrl-0 = <&spi0_pins &spi0_cs_pins>;
-	cs-gpios = <&gpio 8 1>, <&gpio 7 1>;
-
-	spidev0: spidev@0{
-		compatible = "spidev";
-		reg = <0>;	/* CE0 */
-		#address-cells = <1>;
-		#size-cells = <0>;
-		spi-max-frequency = <125000000>;
-	};
-
-	spidev1: spidev@1{
-		compatible = "spidev";
-		reg = <1>;	/* CE1 */
-		#address-cells = <1>;
-		#size-cells = <0>;
-		spi-max-frequency = <125000000>;
-	};
-};
-
 &gpio {
-	spi0_pins: spi0_pins {
-		brcm,pins = <9 10 11>;
-		brcm,function = <BCM2835_FSEL_ALT0>;
-	};
-
-	spi0_cs_pins: spi0_cs_pins {
-		brcm,pins = <8 7>;
-		brcm,function = <BCM2835_FSEL_GPIO_OUT>;
-	};
-
-	spi3_pins: spi3_pins {
-		brcm,pins = <1 2 3>;
-		brcm,function = <BCM2835_FSEL_ALT3>;
-	};
-
-	spi3_cs_pins: spi3_cs_pins {
-		brcm,pins = <0 24>;
-		brcm,function = <BCM2835_FSEL_GPIO_OUT>;
-	};
-
 	spi4_pins: spi4_pins {
 		brcm,pins = <5 6 7>;
 		brcm,function = <BCM2835_FSEL_ALT3>;
@@ -451,38 +384,12 @@
 		brcm,function = <BCM2835_FSEL_GPIO_OUT>;
 	};
 
-	spi5_pins: spi5_pins {
-		brcm,pins = <13 14 15>;
-		brcm,function = <BCM2835_FSEL_ALT3>;
-	};
-
-	spi5_cs_pins: spi5_cs_pins {
-		brcm,pins = <12 26>;
-		brcm,function = <BCM2835_FSEL_GPIO_OUT>;
-	};
-
-	spi6_pins: spi6_pins {
-		brcm,pins = <19 20 21>;
-		brcm,function = <BCM2835_FSEL_ALT3>;
-	};
-
-	spi6_cs_pins: spi6_cs_pins {
-		brcm,pins = <18 27>;
-		brcm,function = <BCM2835_FSEL_GPIO_OUT>;
-	};
-
 	i2c0_pins: i2c0 {
 		brcm,pins = <0 1>;
 		brcm,function = <BCM2835_FSEL_ALT0>;
 		brcm,pull = <BCM2835_PUD_UP>;
 	};
 
-	i2c1_pins: i2c1 {
-		brcm,pins = <2 3>;
-		brcm,function = <BCM2835_FSEL_ALT0>;
-		brcm,pull = <BCM2835_PUD_UP>;
-	};
-
 	i2c3_pins: i2c3 {
 		brcm,pins = <4 5>;
 		brcm,function = <BCM2835_FSEL_ALT5>;
@@ -495,23 +402,6 @@
 		brcm,pull = <BCM2835_PUD_UP>;
 	};
 
-	i2c5_pins: i2c5 {
-		brcm,pins = <12 13>;
-		brcm,function = <BCM2835_FSEL_ALT5>;
-		brcm,pull = <BCM2835_PUD_UP>;
-	};
-
-	i2c6_pins: i2c6 {
-		brcm,pins = <22 23>;
-		brcm,function = <BCM2835_FSEL_ALT5>;
-		brcm,pull = <BCM2835_PUD_UP>;
-	};
-
-	i2s_pins: i2s {
-		brcm,pins = <18 19 20 21>;
-		brcm,function = <BCM2835_FSEL_ALT0>;
-	};
-
 	sdio_pins: sdio_pins {
 		brcm,pins =     <34 35 36 37 38 39>;
 		brcm,function = <BCM2835_FSEL_ALT3>; // alt3 = SD1
@@ -554,29 +444,12 @@
 		brcm,function = <BCM2835_FSEL_ALT4>;
 		brcm,pull = <0 2>;
 	};
-
-	uart5_pins: uart5_pins {
-		brcm,pins = <12 13>;
-		brcm,function = <BCM2835_FSEL_ALT4>;
-		brcm,pull = <0 2>;
-	};
 };
 
 &i2c0if {
 	clock-frequency = <100000>;
 };
 
-&i2c1 {
-	pinctrl-names = "default";
-	pinctrl-0 = <&i2c1_pins>;
-	clock-frequency = <100000>;
-};
-
-&i2s {
-	pinctrl-names = "default";
-	pinctrl-0 = <&i2s_pins>;
-};
-
 // =============================================
 // Board specific stuff here
 
@@ -611,6 +484,81 @@
 		linux,default-trigger = "default-on";
 		gpios = <&expgpio 2 GPIO_ACTIVE_LOW>;
 	};
+
+	wlan2g {
+              label = "wlan2g";
+              gpios = <&gpio 17 GPIO_ACTIVE_HIGH>;
+          };
+  
+      wlan5g {
+              label = "wlan5g";
+              gpios = <&gpio 3 GPIO_ACTIVE_HIGH>;
+          };
+  
+      wan {
+              label = "wan";
+              gpios = <&gpio 27 GPIO_ACTIVE_HIGH>;
+          };  
+     
+      5g1 {
+              label = "5g1";
+              gpios = <&gpio 21 GPIO_ACTIVE_HIGH>;
+          };  
+     
+      5g2 {
+              label = "5g2";
+              gpios = <&gpio 20 GPIO_ACTIVE_HIGH>;
+          };  
+     
+      5g3 {
+              label = "5g3";
+              gpios = <&gpio 16 GPIO_ACTIVE_HIGH>;
+          };  
+  
+      5g4 {
+              label = "5g4";
+              gpios = <&gpio 12 GPIO_ACTIVE_HIGH>;
+          };   
+    
+      5g5 {
+              label = "5g5";
+              gpios = <&gpio 10 GPIO_ACTIVE_HIGH>;
+          }; 
+      
+      5g6 {
+              label = "5g6";
+              gpios = <&gpio 22 GPIO_ACTIVE_HIGH>;
+          };
+    
+      5gpwr1 {
+              label = "5gpwr1";
+              gpios = <&gpio 26 GPIO_ACTIVE_HIGH>;
+          }; 
+      
+      5gpwr2 {
+              label = "5gpwr2";
+              gpios = <&gpio 19 GPIO_ACTIVE_HIGH>;
+          };  
+     
+      5gpwr3 {
+              label = "5gpwr3";
+              gpios = <&gpio 13 GPIO_ACTIVE_HIGH>;
+          };  
+  
+      5gpwr4 {
+              label = "5gpwr4";
+              gpios = <&gpio 24 GPIO_ACTIVE_HIGH>;
+          };    
+   
+      5gpwr5 {
+              label = "5gpwr5";
+              gpios = <&gpio 23 GPIO_ACTIVE_HIGH>;
+          };  
+     
+      5gpwr6 {
+              label = "5gpwr6";
+              gpios = <&gpio 18 GPIO_ACTIVE_HIGH>;
+          };
 };
 
 &pwm1 {
@@ -640,21 +588,6 @@
 		eth_led0 = <&phy1>,"led-modes:0";
 		eth_led1 = <&phy1>,"led-modes:4";
 
-		ant1 =  <&ant1>,"output-high?=on",
-			<&ant1>, "output-low?=off",
-			<&ant2>, "output-high?=off",
-			<&ant2>, "output-low?=on";
-		ant2 =  <&ant1>,"output-high?=off",
-			<&ant1>, "output-low?=on",
-			<&ant2>, "output-high?=on",
-			<&ant2>, "output-low?=off";
-		noant = <&ant1>,"output-high?=off",
-			<&ant1>, "output-low?=on",
-			<&ant2>, "output-high?=off",
-			<&ant2>, "output-low?=on";
-
 		sd_poll_once = <&emmc2>, "non-removable?";
-		spi_dma4 = <&spi0>, "dmas:0=", <&dma40>,
-			   <&spi0>, "dmas:8=", <&dma40>;
 	};
 };
