From 975b265c4a816790df986c9914c72f7258f545fb Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Tue, 20 Feb 2018 20:53:46 +0000
Subject: [PATCH 123/432] hack: cache: Fix linker error

---
 arch/arm/mm/cache-v7.S | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/arm/mm/cache-v7.S b/arch/arm/mm/cache-v7.S
index 7a4c92ec7d91..9370d761a04b 100644
--- a/arch/arm/mm/cache-v7.S
+++ b/arch/arm/mm/cache-v7.S
@@ -350,6 +350,7 @@ ENDPROC(v7_flush_kern_dcache_area)
  *	- start   - virtual start address of region
  *	- end     - virtual end address of region
  */
+ENTRY(b15_dma_inv_range)
 ENTRY(v7_dma_inv_range)
 	dcache_line_size r2, r3
 	sub	r3, r2, #1
@@ -380,6 +381,7 @@ ENDPROC(v7_dma_inv_range)
  *	- start   - virtual start address of region
  *	- end     - virtual end address of region
  */
+ENTRY(b15_dma_clean_range)
 ENTRY(v7_dma_clean_range)
 	dcache_line_size r2, r3
 	sub	r3, r2, #1
-- 
2.19.1

