--- a/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2018-12-26 16:53:06.778501203 +0100
+++ b/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2018-12-26 16:55:46.280213376 +0100
@@ -22,6 +22,7 @@
 	};
 
 	memory {
-		reg = <0 0x80000000 0 0x20000000>;
+		device_type = "memory";
+		reg = <0 0x80000000 0 0x80000000>;
 	};
 
@@ -115,11 +116,6 @@
 		};
 	};
 
-	memory@80000000 {
-		device_type = "memory";
-		reg = <0 0x80000000 0 0x40000000>;
-	};
-
 	mt7530: switch@0 {
 		compatible = "mediatek,mt7530";
 	};
