Index: linux-5.4.164/drivers/mtd/spi-nor/spi-nor.c
===================================================================
--- linux-5.4.164.orig/drivers/mtd/spi-nor/spi-nor.c
+++ linux-5.4.164/drivers/mtd/spi-nor/spi-nor.c
@@ -5117,6 +5117,21 @@ int spi_nor_scan(struct spi_nor *nor, co
 }
 EXPORT_SYMBOL_GPL(spi_nor_scan);
 
+
+//david add flash id to sysfs
+static ssize_t m25p80_fid(struct device *dev, struct device_attribute *attr, char *buf)
+{
+	struct spi_device *spi = to_spi_device(dev);
+	u8 code[5] = {0}, fid[8];
+
+	code[0] = 0x4b;
+	if (spi_write_then_read(spi, &code, 5, fid, 8) < 0)
+		return sprintf(buf, "\n");
+	return sprintf(buf, "%08X%08X\n", ((u32 *)fid)[0], ((u32 *)fid)[1]);
+}
+static DEVICE_ATTR(fid, S_IRUGO, m25p80_fid, NULL);
+//add end
+
 static int spi_nor_probe(struct spi_mem *spimem)
 {
 	struct spi_device *spi = spimem->spi;
@@ -5177,7 +5192,9 @@ static int spi_nor_probe(struct spi_mem
 		if (!nor->bouncebuf)
 			return -ENOMEM;
 	}
-
+//david add flash id to sysfs
+	device_create_file(&spi->dev, &dev_attr_fid);
+//add end
 	return mtd_device_register(&nor->mtd, data ? data->parts : NULL,
 				   data ? data->nr_parts : 0);
 }
