--- a/luci/modules/luci-mod-status/root/usr/share/rpcd/acl.d/luci-mod-status.json	2023-06-29 20:01:01.203922762 +0200
+++ b/luci/modules/luci-mod-status/root/usr/share/rpcd/acl.d/luci-mod-status.json	2023-06-29 20:01:05.479852745 +0200
@@ -33,7 +33,7 @@
 			"file": {
 				"/bin/dmesg -r": [ "exec" ],
 				"/sbin/logread": [ "stat" ],
-				"/sbin/logread -e ^": [ "exec" ],
+				"/sbin/logread": [ "exec" ],
 				"/usr/sbin/logread": [ "stat" ],
 				"/usr/sbin/logread -e ^": [ "exec" ]
 			},
--- a/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/syslog.js	2024-05-14 17:40:42.933677009 +0200
+++ b/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/syslog.js	2024-05-15 13:54:27.933558437 +0200
@@ -12,7 +12,7 @@
 		]).then(function(stat) {
 			var logger = stat[0] ? stat[0].path : stat[1] ? stat[1].path : null;
 
-			return fs.exec_direct(logger, [ '-e', '^' ]).then(logdata => {
+			return fs.exec_direct(logger).then(logdata => {
 				const loglines = logdata.trim().split(/\n/);
 				return { value: loglines.join('\n'), rows: loglines.length + 1 };
 			}).catch(function(err) {
