Index: package/kernel/linux/modules/usb.mk
===================================================================
--- a/package/kernel/linux/modules/usb.mk	(revisione 42462)
+++ b/package/kernel/linux/modules/usb.mk	(copia locale)
@@ -1091,8 +1091,21 @@
 endef
 
 $(eval $(call KernelPackage,usb-net-smsc95xx))
+
+define KernelPackage/usb-net-smsc75xx
+  TITLE:=SMSC LAN75XX based USB 2.0 Gigabit ethernet devices
+  KCONFIG:=CONFIG_USB_NET_SMSC75XX
+  FILES:=$(LINUX_DIR)/drivers/$(USBNET_DIR)/smsc75xx.ko
+  AUTOLOAD:=$(call AutoProbe,smsc75xx)
+  $(call AddDepends/usb-net, +kmod-lib-crc16)
+endef
 
+define KernelPackage/usb-net-smsc75xx/description
+ Kernel module for SMSC LAN75XX based devices
+endef
 
+$(eval $(call KernelPackage,usb-net-smsc75xx))
+
 define KernelPackage/usb-net-dm9601-ether
   TITLE:=Support for DM9601 ethernet connections
   KCONFIG:=CONFIG_USB_NET_DM9601
