Index: linux-5.4.124/arch/arm/boot/dts/qcom-ipq4019.dtsi
===================================================================
--- linux-5.4.124.orig/arch/arm/boot/dts/qcom-ipq4019.dtsi
+++ linux-5.4.124/arch/arm/boot/dts/qcom-ipq4019.dtsi
@@ -140,9 +140,9 @@
 	};
 
 	clocks {
-		sleep_clk: sleep_clk {
+		sleep_clk: gcc_sleep_clk_src {
 			compatible = "fixed-clock";
-			clock-frequency = <32768>;
+			clock-frequency = <32000>;
 			#clock-cells = <0>;
 		};
 
