Index: linux-5.4.147/drivers/mtd/nand/spi/micron.c
===================================================================
--- linux-5.4.147.orig/drivers/mtd/nand/spi/micron.c
+++ linux-5.4.147/drivers/mtd/nand/spi/micron.c
@@ -128,8 +128,8 @@ static const struct spinand_manufacturer
 
 const struct spinand_manufacturer micron_spinand_manufacturer = {
 	.id = SPINAND_MFR_MICRON,
-	.name = "Micron",
+	.name = "Micron / XTX",
 	.chips = micron_spinand_table,
 	.nchips = ARRAY_SIZE(micron_spinand_table),
 	.ops = &micron_spinand_manuf_ops,
 };
