--- a/backport-include/linux/acpi_amd_wbrf.h	2024-09-13 14:55:36.816379630 +0200
+++ b/backport-include/linux/acpi_amd_wbrf.h	2024-09-13 14:55:47.792224160 +0200
@@ -7,9 +7,6 @@
 #ifndef _ACPI_AMD_WBRF_H
 #define _ACPI_AMD_WBRF_H
 
-#if LINUX_VERSION_IS_GEQ(6,8,0)
-#include_next <linux/acpi_amd_wbrf.h>
-#else
 #include <linux/device.h>
 #include <linux/notifier.h>
 
@@ -82,5 +79,4 @@
 	return -ENODEV;
 }
 
-#endif /* >=6,8,0 */
 #endif /* _ACPI_AMD_WBRF_H */
