--- a/target/linux/bcm27xx/modules/sound.mk	2022-01-04 05:35:20.202715290 +0800
+++ b/target/linux/bcm27xx/modules/sound1.mk	2022-01-04 05:35:16.419898622 +0800
@@ -130,29 +130,6 @@
 
 $(eval $(call KernelPackage,sound-soc-allo-boss-dac))
 
-
-define KernelPackage/sound-soc-allo-boss2-dac
-  TITLE:=Support for Allo Boss2 DAC
-  KCONFIG:= \
-    CONFIG_SND_AUDIO_GRAPH_CARD \
-    CONFIG_SND_BCM2708_SOC_ALLO_BOSS2_DAC
-  FILES:= \
-    $(LINUX_DIR)/sound/soc/bcm/snd-soc-allo-boss2-dac.ko
-  AUTOLOAD:=$(call AutoLoad,68,snd-soc-allo-boss2-dac)
-  DEPENDS:= \
-    kmod-sound-soc-bcm2835-i2s \
-    +kmod-i2c-bcm2835 \
-    +kmod-regmap-i2c
-  $(call AddDepends/sound)
-endef
-
-define KernelPackage/sound-soc-allo-boss2-dac/description
-  This package contains support for Allo Boss2 DAC
-endef
-
-$(eval $(call KernelPackage,sound-soc-allo-boss2-dac))
-
-
 define KernelPackage/sound-soc-allo-digione
   TITLE:=Support for Allo Piano DigiOne
   KCONFIG:= \
