--- a/scripts/Makefile	2020-03-21 22:28:13.290800484 +0800
+++ b/scripts/Makefile	2020-03-21 22:28:26.230870790 +0800
@@ -23,8 +23,8 @@ hostprogs-$(CONFIG_SYSTEM_EXTRA_CERTIFIC
 
 HOSTCFLAGS_sortextable.o = -I$(srctree)/tools/include
 HOSTCFLAGS_asn1_compiler.o = -I$(srctree)/include
-HOSTLDLIBS_sign-file = -lcrypto
-HOSTLDLIBS_extract-cert = -lcrypto
+HOSTLDLIBS_sign-file = -lcrypto -lpthread
+HOSTLDLIBS_extract-cert = -lcrypto -lpthread
 
 always		:= $(hostprogs-y) $(hostprogs-m)
 

