--- a/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts.anc	2020-11-16 15:48:00.852410183 +0100
+++ b/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2020-11-16 15:49:23.878965822 +0100
@@ -210,27 +210,27 @@
 
 				port@0 {
 					reg = <0>;
-					label = "wan";
+					label = "lan";
 				};
 
 				port@1 {
 					reg = <1>;
-					label = "lan0";
+					label = "wan1";
 				};
 
 				port@2 {
 					reg = <2>;
-					label = "lan1";
+					label = "wan2";
 				};
 
 				port@3 {
 					reg = <3>;
-					label = "lan2";
+					label = "wan3";
 				};
 
 				port@4 {
 					reg = <4>;
-					label = "lan3";
+					label = "wan4";
 				};
 
 				port@6 {
