<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
--- a/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2020-04-19 11:02:56.505715879 +0200
+++ b/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2020-04-19 11:03:53.620780390 +0200
@@ -196,7 +196,7 @@
 	gmac1: mac@1 {
 		compatible = "mediatek,eth-mac";
 		reg = <1>;
-		label = "wan";
+		label = "lan";
 		phy-mode = "rgmii";
 		phy-handle = <&ephy0>;
 		local-mac-address = [00 0a 35 00 00 02];
@@ -221,31 +221,31 @@
 				#address-cells = <1>;
 				#size-cells = <0>;
 
-/* Disabled, is now handled by gmac1 (eth1/wan) via phy-handle!
+/* Disabled, is now handled by gmac1 (eth1/lan) via phy-handle!
=======
=======
>>>>>>> parent of 602cd0b... fixomr
=======
>>>>>>> parent of 602cd0b... fixomr
--- a/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts.anc	2020-11-16 15:48:00.852410183 +0100
+++ b/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2020-11-16 15:49:23.878965822 +0100
@@ -210,27 +210,27 @@
 
<<<<<<< HEAD
<<<<<<< HEAD
>>>>>>> parent of 602cd0b... fixomr
=======
>>>>>>> parent of 602cd0b... fixomr
=======
>>>>>>> parent of 602cd0b... fixomr
 				port@0 {
 					reg = <0>;
-					label = "wan";
+					label = "lan";
 				};
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
 */
=======
>>>>>>> parent of 602cd0b... fixomr
=======
>>>>>>> parent of 602cd0b... fixomr
=======
>>>>>>> parent of 602cd0b... fixomr
 
 				port@1 {
 					reg = <1>;
-					label = "lan0";
+					label = "wan1";
 				};
 
 				port@2 {
 					reg = <2>;
-					label = "lan1";
+					label = "wan2";
 				};
 
 				port@3 {
 					reg = <3>;
-					label = "lan2";
+					label = "wan3";
 				};
 
 				port@4 {
 					reg = <4>;
-					label = "lan3";
+					label = "wan4";
 				};
 
 				port@6 {
