--- a/hif/pcie/pcie.c	2021-11-26 16:47:58.360969448 +0100
+++ b/hif/pcie/pcie.c	2021-11-26 16:48:46.108695785 +0100
@@ -1293,8 +1293,10 @@
 	char *buf = &str_buf[0];
 	mm_segment_t oldfs;
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0))
 	oldfs = get_fs();
 	set_fs(KERNEL_DS);
+#endif
 
 	buf += sprintf(buf, "\nMAC: %pM\n", bf_mimo_ctrl->rec_mac);
 	buf += sprintf(buf, "SU_0_MU_1: %d\n", bf_mimo_ctrl->type);
@@ -1314,7 +1316,9 @@
 			  filename, (unsigned int)fp_data);
 	}
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0))
 	set_fs(oldfs);
+#endif
 }
 
 static void pcie_process_account(struct ieee80211_hw *hw)
