From 3332a6450700b6a1897547fec40248155a6b6459 Mon Sep 17 00:00:00 2001
From: Jacko Dirks <jdirks.linuxdev@gmail.com>
Date: Tue, 5 May 2020 14:28:14 +0200
Subject: [PATCH 240/726] media: i2c: tc358743: Fix fallthrough warning

Signed-off-by: Jacko Dirks <jdirks.linuxdev@gmail.com>
---
 drivers/media/i2c/tc358743.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/media/i2c/tc358743.c b/drivers/media/i2c/tc358743.c
index 69127833f4c7..618fbdc51794 100644
--- a/drivers/media/i2c/tc358743.c
+++ b/drivers/media/i2c/tc358743.c
@@ -1983,6 +1983,7 @@ static int tc358743_probe_of(struct tc358743_state *state)
 	switch (bps_pr_lane) {
 	default:
 		dev_warn(dev, "untested bps per lane: %u bps\n", bps_pr_lane);
+		fallthrough;
 	case 594000000U:
 		state->pdata.lineinitcnt = 0xe80;
 		state->pdata.lptxtimecnt = 0x003;
-- 
2.33.1

