diff --git a/drivers/net/usb/lan78xx.c b/drivers/net/usb/lan78xx.c
index 8761bc7..c0d30e9 100644
--- a/drivers/net/usb/lan78xx.c
+++ b/drivers/net/usb/lan78xx.c
@@ -1243,7 +1243,11 @@ static void lan78xx_defer_kevent(struct lan78xx_net *dev, int work)
 {
 	set_bit(work, &dev->flags);
 	if (!schedule_delayed_work(&dev->wq, 0))
-		netdev_err(dev->net, "kevent %d may have been dropped\n", work);
+// XXX : LBO patch - this mitigates (but doesn't fix) issue #2164
+//		netdev_err(dev->net, "kevent %d may have been dropped\n", work);
+		{};
+// End of Patch
+
 }
 
 static void lan78xx_status(struct lan78xx_net *dev, struct urb *urb)