From e13668fdd0f92f9346543d1f3eba2e857e8d9c66 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Thu, 3 Sep 2020 14:02:21 +0100
Subject: [PATCH 269/552] bcm2835-mmc: uninitialized_var is no more

---
 drivers/mmc/host/bcm2835-mmc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/mmc/host/bcm2835-mmc.c b/drivers/mmc/host/bcm2835-mmc.c
index caf46be8eb0d..cdc1318e5936 100644
--- a/drivers/mmc/host/bcm2835-mmc.c
+++ b/drivers/mmc/host/bcm2835-mmc.c
@@ -367,7 +367,7 @@ static void bcm2835_bcm2835_mmc_read_block_pio(struct bcm2835_host *host)
 	unsigned long flags;
 	size_t blksize, len, chunk;
 
-	u32 uninitialized_var(scratch);
+	u32 scratch = 0;
 	u8 *buf;
 
 	blksize = host->data->blksz;
-- 
2.33.0

