--- a/nss_dp_main.c	2023-04-19 13:59:34.762040822 +0200
+++ b/nss_dp_main.c	2023-04-19 13:59:50.057767685 +0200
@@ -601,7 +601,7 @@
 	if (!ret && is_valid_ether_addr(maddr)) {
 		ether_addr_copy(netdev->dev_addr, maddr);
 	} else {
-		random_ether_addr(netdev->dev_addr);
+		eth_random_addr(netdev->dev_addr);
 		pr_info("GMAC%d(%px) Invalid MAC@ - using %pM\n", dp_priv->macid,
 						dp_priv, netdev->dev_addr);
 	}
