--- a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c	2023-04-19 14:29:23.502166413 +0200
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c	2023-04-19 14:29:35.865951285 +0200
@@ -838,8 +838,7 @@
 	 * NAPI add
 	 */
 	if (!edma_hw.napi_added) {
-		netif_napi_add(netdev, &edma_hw.napi, edma_napi,
-				NAPI_POLL_WEIGHT);
+		netif_napi_add(netdev, &edma_hw.napi, edma_napi);
 		/*
 		 * Register the interrupt handlers and enable interrupts
 		 */
