--- a/drivers/net/wireless/ath/ath10k/pci.c	2023-12-22 19:22:13.061853464 +0100
+++ b/drivers/net/wireless/ath/ath10k/pci.c	2023-12-22 19:23:22.184610463 +0100
@@ -3163,7 +3163,11 @@
 		 * immediate servicing.
 		 */
 		if (ath10k_ce_interrupt_summary(ar)) {
+#if LINUX_VERSION_IS_LESS(6,7,0)
 			napi_reschedule(ctx);
+#else
+			napi_schedule(ctx);
+#endif
 			goto out;
 		}
 		ath10k_pci_enable_legacy_irq(ar);
