--- a/drivers/net/wireless/intel/iwlwifi/mvm/tt.c	2024-07-03 19:59:46.724206025 +0200
+++ b/drivers/net/wireless/intel/iwlwifi/mvm/tt.c	2024-07-03 19:59:56.696041840 +0200
@@ -696,7 +696,6 @@
 	mvm->tz_device.tzone = thermal_zone_device_register_with_trips(name,
 							mvm->tz_device.trips,
 							IWL_MAX_DTS_TRIPS,
-							IWL_WRITABLE_TRIPS_MSK,
 							mvm, &tzone_ops,
 							NULL, 0, 0);
 	if (IS_ERR(mvm->tz_device.tzone)) {
