diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/admtek/adm8211.c backports-5.10.42-1/drivers/net/wireless/admtek/adm8211.c
--- backports-5.10.42-1.anc/drivers/net/wireless/admtek/adm8211.c	2021-09-23 22:26:42.899395773 +0200
+++ backports-5.10.42-1/drivers/net/wireless/admtek/adm8211.c	2021-09-23 22:30:53.147231692 +0200
@@ -28,7 +28,6 @@
 MODULE_AUTHOR("Michael Wu <flamingice@sourmilk.net>");
 MODULE_AUTHOR("Jouni Malinen <j@w1.fi>");
 MODULE_DESCRIPTION("Driver for IEEE 802.11b wireless cards based on ADMtek ADM8211");
-MODULE_SUPPORTED_DEVICE("ADM8211");
 MODULE_LICENSE("GPL");
 
 static unsigned int tx_ring_size __read_mostly = 16;
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ath/ath5k/base.c backports-5.10.42-1/drivers/net/wireless/ath/ath5k/base.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ath/ath5k/base.c	2021-09-23 22:26:42.903395707 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ath/ath5k/base.c	2021-09-23 22:30:46.011350432 +0200
@@ -90,7 +90,6 @@
 MODULE_AUTHOR("Jiri Slaby");
 MODULE_AUTHOR("Nick Kossifidis");
 MODULE_DESCRIPTION("Support for 5xxx series of Atheros 802.11 wireless LAN cards.");
-MODULE_SUPPORTED_DEVICE("Atheros 5xxx WLAN cards");
 MODULE_LICENSE("Dual BSD/GPL");
 
 static int ath5k_init(struct ieee80211_hw *hw);
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ath/ath9k/hw.c backports-5.10.42-1/drivers/net/wireless/ath/ath9k/hw.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ath/ath9k/hw.c	2021-09-23 22:26:42.907395640 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ath/ath9k/hw.c	2021-09-23 22:30:44.355377987 +0200
@@ -34,7 +34,6 @@
 
 MODULE_AUTHOR("Atheros Communications");
 MODULE_DESCRIPTION("Support for Atheros 802.11n wireless LAN cards.");
-MODULE_SUPPORTED_DEVICE("Atheros 802.11n WLAN cards");
 MODULE_LICENSE("Dual BSD/GPL");
 
 static void ath9k_hw_set_clockrate(struct ath_hw *ah)
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ath/ath9k/init.c backports-5.10.42-1/drivers/net/wireless/ath/ath9k/init.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ath/ath9k/init.c	2021-09-23 22:26:42.907395640 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ath/ath9k/init.c	2021-09-23 22:30:44.355377987 +0200
@@ -37,7 +37,6 @@
 
 MODULE_AUTHOR("Atheros Communications");
 MODULE_DESCRIPTION("Support for Atheros 802.11n wireless LAN cards.");
-MODULE_SUPPORTED_DEVICE("Atheros 802.11n WLAN cards");
 MODULE_LICENSE("Dual BSD/GPL");
 
 static unsigned int ath9k_debug = ATH_DBG_DEFAULT;
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/atmel/atmel.c backports-5.10.42-1/drivers/net/wireless/atmel/atmel.c
--- backports-5.10.42-1.anc/drivers/net/wireless/atmel/atmel.c	2021-09-23 22:26:42.911395574 +0200
+++ backports-5.10.42-1/drivers/net/wireless/atmel/atmel.c	2021-09-23 22:30:46.023350232 +0200
@@ -75,7 +75,6 @@
 MODULE_AUTHOR("Simon Kelley");
 MODULE_DESCRIPTION("Support for Atmel at76c50x 802.11 wireless ethernet cards.");
 MODULE_LICENSE("GPL");
-MODULE_SUPPORTED_DEVICE("Atmel at76c50x wireless cards");
 
 /* The name of the firmware file to be loaded
    over-rides any automatic selection */
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/atmel/atmel_cs.c backports-5.10.42-1/drivers/net/wireless/atmel/atmel_cs.c
--- backports-5.10.42-1.anc/drivers/net/wireless/atmel/atmel_cs.c	2021-09-23 22:26:42.911395574 +0200
+++ backports-5.10.42-1/drivers/net/wireless/atmel/atmel_cs.c	2021-09-23 22:30:46.023350232 +0200
@@ -57,7 +57,6 @@
 MODULE_AUTHOR("Simon Kelley");
 MODULE_DESCRIPTION("Support for Atmel at76c50x 802.11 wireless ethernet cards.");
 MODULE_LICENSE("GPL");
-MODULE_SUPPORTED_DEVICE("Atmel at76c50x PCMCIA cards");
 
 /*====================================================================*/
 
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/atmel/atmel_pci.c backports-5.10.42-1/drivers/net/wireless/atmel/atmel_pci.c
--- backports-5.10.42-1.anc/drivers/net/wireless/atmel/atmel_pci.c	2021-09-23 22:26:42.911395574 +0200
+++ backports-5.10.42-1/drivers/net/wireless/atmel/atmel_pci.c	2021-09-23 22:30:46.023350232 +0200
@@ -16,7 +16,6 @@
 MODULE_AUTHOR("Simon Kelley");
 MODULE_DESCRIPTION("Support for Atmel at76c50x 802.11 wireless ethernet cards.");
 MODULE_LICENSE("GPL");
-MODULE_SUPPORTED_DEVICE("Atmel at76c506 PCI wireless cards");
 
 static const struct pci_device_id card_ids[] = {
 	{ 0x1114, 0x0506, PCI_ANY_ID, PCI_ANY_ID },
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/broadcom/brcm80211/brcmsmac/mac80211_if.c backports-5.10.42-1/drivers/net/wireless/broadcom/brcm80211/brcmsmac/mac80211_if.c
--- backports-5.10.42-1.anc/drivers/net/wireless/broadcom/brcm80211/brcmsmac/mac80211_if.c	2021-09-23 22:26:42.915395506 +0200
+++ backports-5.10.42-1/drivers/net/wireless/broadcom/brcm80211/brcmsmac/mac80211_if.c	2021-09-23 22:30:48.843303308 +0200
@@ -87,7 +87,6 @@
 
 MODULE_AUTHOR("Broadcom Corporation");
 MODULE_DESCRIPTION("Broadcom 802.11n wireless LAN driver.");
-MODULE_SUPPORTED_DEVICE("Broadcom 802.11n WLAN cards");
 MODULE_LICENSE("Dual BSD/GPL");
 /* This needs to be adjusted when brcms_firmwares changes */
 MODULE_FIRMWARE("brcm/bcm43xx-0.fw");
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/broadcom/brcm80211/brcmutil/utils.c backports-5.10.42-1/drivers/net/wireless/broadcom/brcm80211/brcmutil/utils.c
--- backports-5.10.42-1.anc/drivers/net/wireless/broadcom/brcm80211/brcmutil/utils.c	2021-09-23 22:26:42.915395506 +0200
+++ backports-5.10.42-1/drivers/net/wireless/broadcom/brcm80211/brcmutil/utils.c	2021-09-23 22:30:51.611257250 +0200
@@ -12,7 +12,6 @@
 
 MODULE_AUTHOR("Broadcom Corporation");
 MODULE_DESCRIPTION("Broadcom 802.11n wireless LAN driver utilities.");
-MODULE_SUPPORTED_DEVICE("Broadcom 802.11n WLAN cards");
 MODULE_LICENSE("Dual BSD/GPL");
 
 struct sk_buff *brcmu_pkt_buf_get_skb(uint len)
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/cisco/airo.c backports-5.10.42-1/drivers/net/wireless/cisco/airo.c
--- backports-5.10.42-1.anc/drivers/net/wireless/cisco/airo.c	2021-09-23 22:26:42.915395506 +0200
+++ backports-5.10.42-1/drivers/net/wireless/cisco/airo.c	2021-09-23 22:30:53.147231692 +0200
@@ -251,7 +251,6 @@
 MODULE_DESCRIPTION("Support for Cisco/Aironet 802.11 wireless ethernet cards.  "
 		   "Direct support for ISA/PCI/MPI cards and support for PCMCIA when used with airo_cs.");
 MODULE_LICENSE("Dual BSD/GPL");
-MODULE_SUPPORTED_DEVICE("Aironet 4500, 4800 and Cisco 340/350");
 module_param_hw_array(io, int, ioport, NULL, 0);
 module_param_hw_array(irq, int, irq, NULL, 0);
 module_param_array(rates, int, NULL, 0);
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/cisco/airo_cs.c backports-5.10.42-1/drivers/net/wireless/cisco/airo_cs.c
--- backports-5.10.42-1.anc/drivers/net/wireless/cisco/airo_cs.c	2021-09-23 22:26:42.915395506 +0200
+++ backports-5.10.42-1/drivers/net/wireless/cisco/airo_cs.c	2021-09-23 22:30:53.159231492 +0200
@@ -47,7 +47,6 @@
 		   "cards.  This is the module that links the PCMCIA card "
 		   "with the airo module.");
 MODULE_LICENSE("Dual BSD/GPL");
-MODULE_SUPPORTED_DEVICE("Aironet 4500, 4800 and Cisco 340 PCMCIA cards");
 
 /*====================================================================*/
 
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/intersil/hostap/hostap_cs.c backports-5.10.42-1/drivers/net/wireless/intersil/hostap/hostap_cs.c
--- backports-5.10.42-1.anc/drivers/net/wireless/intersil/hostap/hostap_cs.c	2021-09-23 22:26:42.919395441 +0200
+++ backports-5.10.42-1/drivers/net/wireless/intersil/hostap/hostap_cs.c	2021-09-23 22:30:53.027233689 +0200
@@ -26,7 +26,6 @@
 MODULE_AUTHOR("Jouni Malinen");
 MODULE_DESCRIPTION("Support for Intersil Prism2-based 802.11 wireless LAN "
 		   "cards (PC Card).");
-MODULE_SUPPORTED_DEVICE("Intersil Prism2-based WLAN cards (PC Card)");
 MODULE_LICENSE("GPL");
 
 
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/intersil/hostap/hostap_pci.c backports-5.10.42-1/drivers/net/wireless/intersil/hostap/hostap_pci.c
--- backports-5.10.42-1.anc/drivers/net/wireless/intersil/hostap/hostap_pci.c	2021-09-23 22:26:42.919395441 +0200
+++ backports-5.10.42-1/drivers/net/wireless/intersil/hostap/hostap_pci.c	2021-09-23 22:30:53.027233689 +0200
@@ -27,7 +27,6 @@
 MODULE_AUTHOR("Jouni Malinen");
 MODULE_DESCRIPTION("Support for Intersil Prism2.5-based 802.11 wireless LAN "
 		   "PCI cards.");
-MODULE_SUPPORTED_DEVICE("Intersil Prism2.5-based WLAN PCI cards");
 MODULE_LICENSE("GPL");
 
 
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/intersil/hostap/hostap_plx.c backports-5.10.42-1/drivers/net/wireless/intersil/hostap/hostap_plx.c
--- backports-5.10.42-1.anc/drivers/net/wireless/intersil/hostap/hostap_plx.c	2021-09-23 22:26:42.919395441 +0200
+++ backports-5.10.42-1/drivers/net/wireless/intersil/hostap/hostap_plx.c	2021-09-23 22:30:53.031233622 +0200
@@ -30,7 +30,6 @@
 MODULE_AUTHOR("Jouni Malinen");
 MODULE_DESCRIPTION("Support for Intersil Prism2-based 802.11 wireless LAN "
 		   "cards (PLX).");
-MODULE_SUPPORTED_DEVICE("Intersil Prism2-based WLAN cards (PLX)");
 MODULE_LICENSE("GPL");
 
 
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2400pci.c backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2400pci.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2400pci.c	2021-09-23 22:26:42.927395307 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2400pci.c	2021-09-23 22:30:53.039233489 +0200
@@ -1821,7 +1821,6 @@
 MODULE_AUTHOR(DRV_PROJECT);
 MODULE_VERSION(DRV_VERSION);
 MODULE_DESCRIPTION("Ralink RT2400 PCI & PCMCIA Wireless LAN driver.");
-MODULE_SUPPORTED_DEVICE("Ralink RT2460 PCI & PCMCIA chipset based cards");
 MODULE_DEVICE_TABLE(pci, rt2400pci_device_table);
 MODULE_LICENSE("GPL");
 
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2500pci.c backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2500pci.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2500pci.c	2021-09-23 22:26:42.927395307 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2500pci.c	2021-09-23 22:30:53.039233489 +0200
@@ -2119,7 +2119,6 @@
 MODULE_AUTHOR(DRV_PROJECT);
 MODULE_VERSION(DRV_VERSION);
 MODULE_DESCRIPTION("Ralink RT2500 PCI & PCMCIA Wireless LAN driver.");
-MODULE_SUPPORTED_DEVICE("Ralink RT2560 PCI & PCMCIA chipset based cards");
 MODULE_DEVICE_TABLE(pci, rt2500pci_device_table);
 MODULE_LICENSE("GPL");
 
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2500usb.c backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2500usb.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2500usb.c	2021-09-23 22:26:42.927395307 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2500usb.c	2021-09-23 22:30:53.039233489 +0200
@@ -1956,7 +1956,6 @@
 MODULE_AUTHOR(DRV_PROJECT);
 MODULE_VERSION(DRV_VERSION);
 MODULE_DESCRIPTION("Ralink RT2500 USB Wireless LAN driver.");
-MODULE_SUPPORTED_DEVICE("Ralink RT2570 USB chipset based cards");
 MODULE_DEVICE_TABLE(usb, rt2500usb_device_table);
 MODULE_LICENSE("GPL");
 
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2800pci.c backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2800pci.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2800pci.c	2021-09-23 22:26:42.927395307 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2800pci.c	2021-09-23 22:30:53.039233489 +0200
@@ -447,7 +447,6 @@
 MODULE_AUTHOR(DRV_PROJECT);
 MODULE_VERSION(DRV_VERSION);
 MODULE_DESCRIPTION("Ralink RT2800 PCI & PCMCIA Wireless LAN driver.");
-MODULE_SUPPORTED_DEVICE("Ralink RT2860 PCI & PCMCIA chipset based cards");
 MODULE_FIRMWARE(FIRMWARE_RT2860);
 MODULE_DEVICE_TABLE(pci, rt2800pci_device_table);
 MODULE_LICENSE("GPL");
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2800usb.c backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2800usb.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt2800usb.c	2021-09-23 22:26:42.927395307 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt2800usb.c	2021-09-23 22:30:53.039233489 +0200
@@ -1254,7 +1254,6 @@
 MODULE_AUTHOR(DRV_PROJECT);
 MODULE_VERSION(DRV_VERSION);
 MODULE_DESCRIPTION("Ralink RT2800 USB Wireless LAN driver.");
-MODULE_SUPPORTED_DEVICE("Ralink RT2870 USB chipset based cards");
 MODULE_DEVICE_TABLE(usb, rt2800usb_device_table);
 MODULE_FIRMWARE(FIRMWARE_RT2870);
 MODULE_LICENSE("GPL");
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt61pci.c backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt61pci.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt61pci.c	2021-09-23 22:26:42.927395307 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt61pci.c	2021-09-23 22:30:53.047233356 +0200
@@ -2993,7 +2993,5 @@
 MODULE_AUTHOR(DRV_PROJECT);
 MODULE_VERSION(DRV_VERSION);
 MODULE_DESCRIPTION("Ralink RT61 PCI & PCMCIA Wireless LAN driver.");
-MODULE_SUPPORTED_DEVICE("Ralink RT2561, RT2561s & RT2661 "
-			"PCI & PCMCIA chipset based cards");
 MODULE_DEVICE_TABLE(pci, rt61pci_device_table);
 MODULE_FIRMWARE(FIRMWARE_RT2561);
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt73usb.c backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt73usb.c
--- backports-5.10.42-1.anc/drivers/net/wireless/ralink/rt2x00/rt73usb.c	2021-09-23 22:26:42.927395307 +0200
+++ backports-5.10.42-1/drivers/net/wireless/ralink/rt2x00/rt73usb.c	2021-09-23 22:30:53.043233423 +0200
@@ -2513,7 +2513,6 @@
 MODULE_AUTHOR(DRV_PROJECT);
 MODULE_VERSION(DRV_VERSION);
 MODULE_DESCRIPTION("Ralink RT73 USB Wireless LAN driver.");
-MODULE_SUPPORTED_DEVICE("Ralink RT2571W & RT2671 USB chipset based cards");
 MODULE_DEVICE_TABLE(usb, rt73usb_device_table);
 MODULE_FIRMWARE(FIRMWARE_RT2571);
 MODULE_LICENSE("GPL");
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/rsi/rsi_91x_main.c backports-5.10.42-1/drivers/net/wireless/rsi/rsi_91x_main.c
--- backports-5.10.42-1.anc/drivers/net/wireless/rsi/rsi_91x_main.c	2021-09-23 22:26:42.935395174 +0200
+++ backports-5.10.42-1/drivers/net/wireless/rsi/rsi_91x_main.c	2021-09-23 22:30:53.059233156 +0200
@@ -441,6 +441,5 @@
 module_exit(rsi_91x_hal_module_exit);
 MODULE_AUTHOR("Redpine Signals Inc");
 MODULE_DESCRIPTION("Station driver for RSI 91x devices");
-MODULE_SUPPORTED_DEVICE("RSI-91x");
 MODULE_VERSION("0.1");
 MODULE_LICENSE("Dual BSD/GPL");
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/rsi/rsi_91x_sdio.c backports-5.10.42-1/drivers/net/wireless/rsi/rsi_91x_sdio.c
--- backports-5.10.42-1.anc/drivers/net/wireless/rsi/rsi_91x_sdio.c	2021-09-23 22:26:42.935395174 +0200
+++ backports-5.10.42-1/drivers/net/wireless/rsi/rsi_91x_sdio.c	2021-09-23 22:30:53.063233090 +0200
@@ -1571,7 +1571,6 @@
 
 MODULE_AUTHOR("Redpine Signals Inc");
 MODULE_DESCRIPTION("Common SDIO layer for RSI drivers");
-MODULE_SUPPORTED_DEVICE("RSI-91x");
 MODULE_DEVICE_TABLE(sdio, rsi_dev_table);
 MODULE_FIRMWARE(FIRMWARE_RSI9113);
 MODULE_VERSION("0.1");
diff -aurN backports-5.10.42-1.anc/drivers/net/wireless/rsi/rsi_91x_usb.c backports-5.10.42-1/drivers/net/wireless/rsi/rsi_91x_usb.c
--- backports-5.10.42-1.anc/drivers/net/wireless/rsi/rsi_91x_usb.c	2021-09-23 22:26:42.935395174 +0200
+++ backports-5.10.42-1/drivers/net/wireless/rsi/rsi_91x_usb.c	2021-09-23 22:30:53.063233090 +0200
@@ -926,7 +926,6 @@
 
 MODULE_AUTHOR("Redpine Signals Inc");
 MODULE_DESCRIPTION("Common USB layer for RSI drivers");
-MODULE_SUPPORTED_DEVICE("RSI-91x");
 MODULE_DEVICE_TABLE(usb, rsi_dev_table);
 MODULE_FIRMWARE(FIRMWARE_RSI9113);
 MODULE_VERSION("0.1");
