--- a/nat46/modules/nat46-netdev.c	2024-09-13 09:39:48.960815094 +0200
+++ b/nat46/modules/nat46-netdev.c	2024-09-13 09:40:11.128500472 +0200
@@ -110,7 +110,6 @@
 	dev->hard_header_len = 0;
 	dev->addr_len = 0;
 	dev->mtu = 16384; /* iptables does reassembly. Rather than using ETH_DATA_LEN, let's try to get as much mileage as we can with the Linux stack */
-	dev->features = NETIF_F_NETNS_LOCAL;
 	dev->flags = IFF_NOARP | IFF_POINTOPOINT;
 }
 
