--- a/drivers/net/wireless/ath/carl9170/tx.c	2023-06-27 09:18:10.616850964 +0200
+++ b/drivers/net/wireless/ath/carl9170/tx.c	2023-06-27 09:18:39.260374789 +0200
@@ -280,7 +280,8 @@
 	 * carl9170_tx_fill_rateinfo() has filled the rate information
 	 * before we get to this point.
 	 */
-	memset_after(&txinfo->status, 0, rates);
+	memset(&txinfo->pad, 0, sizeof(txinfo->pad));
+	memset(&txinfo->rate_driver_data, 0, sizeof(txinfo->rate_driver_data));
 
 	if (atomic_read(&ar->tx_total_queued))
 		ar->tx_schedule = true;
