Index: linux-5.4.147/drivers/usb/host/xhci.h
===================================================================
--- linux-5.4.147.orig/drivers/usb/host/xhci.h
+++ linux-5.4.147/drivers/usb/host/xhci.h
@@ -1654,7 +1654,7 @@ struct urb_priv {
  * (1K bytes * 8bytes/bit) / (4*32 bits) = 64 segment entries in the table,
  * meaning 64 ring segments.
  * Initial allocated size of the ERST, in number of entries */
-#define	ERST_NUM_SEGS	1
+#define	ERST_NUM_SEGS	16
 /* Initial allocated size of the ERST, in number of entries */
 #define	ERST_SIZE	64
 /* Initial number of event segment rings allocated */
