--- a/src/linux/ifxos_linux_thread_drv.c	2023-05-22 08:50:14.087507831 +0200
+++ b/src/linux/ifxos_linux_thread_drv.c	2023-05-22 08:51:32.702129369 +0200
@@ -154,8 +154,11 @@
    retVal = pThrCntrl->pThrFct(&pThrCntrl->thrParams);
    pThrCntrl->thrParams.bRunning = IFX_FALSE;
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5,17,0))
    complete_and_exit(&pThrCntrl->thrCompletion, (long)retVal);
-
+#else
+   kthread_complete_and_exit(&pThrCntrl->thrCompletion, (long)retVal);
+#endif
    IFXOS_PRN_USR_DBG_NL( IFXOS, IFXOS_PRN_LEVEL_NORMAL,
       ("EXIT - Kernel Thread Startup <%s>" IFXOS_CRLF,
         pThrCntrl->thrParams.pName));
